### PR TITLE
(Core) Forward core changes

### PR DESCRIPTION
**Pull request by acidmanifesto**
This is related to a exsisting pr i am doing. PR https://github.com/azerothcore/azerothcore-wotlk/pull/8760
As I am still trying to identify the underlying issue with auras and conditions, I did fix and refactored a few things into the core on other issues to help it move forward.  This is to help[ with future refactoring while maintain what we have in AZ core with out currently breaking anything. Hopefully. Let the compile testing odds forever be in our favor.

## Changes Proposed:
-  https://github.com/TrinityCore/TrinityCore/commit/449ec0d6ff74ccd1874bc09073353d97c1d0151c
-  https://github.com/TrinityCore/TrinityCore/commit/8775f8b28a5ad4403371577787131ff81f14332b
- Fixed Spell Click Conditions

## Issues Addressed:
- Moving the core forward.

## SOURCE:
Trinitycore changes listed above.

## Tests Performed:
Enables: 
Allow arbitrary containers in grid searchers that support push_back without breaking existing grid searchers in play.
* Makes it possible to use vector and dequeue with grid searchers.

Don't ignore SPELL_ATTR3_ONLY_TARGET_PLAYERS in area auras.

Fixed Spell Click Conditions 

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -4096,6 +4100,10 @@ void Unit::_UnapplyAura(AuraApplicationMap::iterator& i, AuraRemoveMode removeMo
 
     aura->HandleAuraSpecificMods(aurApp, caster, false, false);
 
+        if (Player* player = ToPlayer())
+        if (sConditionMgr->IsSpellUsedInSpellClickConditions(aurApp->GetBase()->GetId()))
+            player->UpdateVisibleGameobjectsOrSpellClicks();
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:4105]: brackets + indentation

Comment (acidmanifesto) [reply src/server/game/Entities/Unit/Unit.cpp:4107]: {
    done;
}

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -4096,6 +4102,10 @@ void Unit::_UnapplyAura(AuraApplicationMap::iterator& i, AuraRemoveMode removeMo
...
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.cpp:4107]: indentation + brackets

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -4018,6 +4018,12 @@ void Unit::_ApplyAura(AuraApplication* aurApp, uint8 effMask)
         if (effMask & 1 << i && (!aurApp->GetRemoveMode()))
             aurApp->_HandleEffect(i, true);
     }
+
+    if (Player* player = ToPlayer())
+        if (sConditionMgr->IsSpellUsedInSpellClickConditions(aurApp->GetBase()->GetId()))
+		{
+			player->UpdateVisibleGameobjectsOrSpellClicks();
+		}
```
[the remark this answers is not in the document]
Comment (Kitzunu) [reply src/server/game/Entities/Unit/Unit.cpp:4026]: Dont use tabs. Use 4 sapces

File: src/server/game/Entities/Unit/Unit.cpp (hunk 2 above)
Comment (acidmanifesto) [reply src/server/game/Entities/Unit/Unit.cpp]: done

File: src/server/game/Entities/Unit/Unit.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (acidmanifesto) [reply src/server/game/Entities/Unit/Unit.cpp:4026]: ok. this is gonna have to be a web edit as it isnt showing that on my end. standby

**Comment by acidmanifesto** [comment]
Docker hates me :(

File: src/server/game/Entities/Unit/Unit.cpp [hunk 4]
```diff
@@ -4096,6 +4104,14 @@ void Unit::_UnapplyAura(AuraApplicationMap::iterator& i, AuraRemoveMode removeMo
 
     aura->HandleAuraSpecificMods(aurApp, caster, false, false);
 
+        if (Player* player = ToPlayer())
+        {
+            if (sConditionMgr->IsSpellUsedInSpellClickConditions(aurApp->GetBase()->GetId()))
+            {
+            player->UpdateVisibleGameobjectsOrSpellClicks();
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Unit/Unit.cpp:4111]: indentation

Comment (acidmanifesto) [reply src/server/game/Entities/Unit/Unit.cpp]: done

Comment (Winfidonarleyan) [reply src/server/game/Entities/Unit/Unit.cpp]: Best

**Review by acidmanifesto — COMMENTED** [review]
Done

**Review by Nyeriah — APPROVED** [review]

**Comment by Nyeriah** [comment]
@azerothcore/testers could someone test this one please?

**Review by Kitzunu — APPROVED** [review]

File: src/server/game/Grids/Notifiers/GridNotifiers.h [hunk 5]
```diff
@@ -800,20 +823,22 @@ namespace Acore
     {
     public:
         AnyUnfriendlyUnitInObjectRangeCheck(WorldObject const* obj, Unit const* funit, float range) : i_obj(obj), i_funit(funit), i_range(range) {}
-        bool operator()(Unit* u)
+
+        bool operator()(Unit* u) const
         {
-            if (u->IsAlive() && i_obj->IsWithinDistInMap(u, i_range) && !i_funit->IsFriendlyTo(u) &&
-                    (i_funit->GetTypeId() != TYPEID_UNIT || !i_funit->ToCreature()->IsAvoidingAOE())) // pussywizard
```
Comment (UltraNix) [inline src/server/game/Grids/Notifiers/GridNotifiers.h:806]: Why removed that?

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 6]
```diff
@@ -2785,43 +2798,37 @@ void DynObjAura::FillTargetMap(std::map<Unit*, uint8>& targets, Unit* /*caster*/
...
             Acore::AnyAoETargetUnitInObjectRangeCheck u_check(GetDynobjOwner(), dynObjOwnerCaster, radius);
-            Acore::UnitListSearcher<Acore::AnyAoETargetUnitInObjectRangeCheck> searcher(GetDynobjOwner(), targetList, u_check);
+            Acore::UnitListSearcher<Acore::AnyAoETargetUnitInObjectRangeCheck> searcher(GetDynobjOwner(), units, u_check);
             Cell::VisitAllObjects(GetDynobjOwner(), searcher, radius);
         }
 
-        for (UnitList::iterator itr = targetList.begin(); itr != targetList.end(); ++itr)
+        for (Unit* unit : units)
         {
-            // xinef: check z level and los dependence
```
Comment (UltraNix) [inline src/server/game/Spells/Auras/SpellAuras.cpp:2813]: Why removed that block?

File: src/server/scripts/Pet/pet_dk.cpp [hunk 7]
```diff
@@ -168,9 +168,9 @@ class npc_pet_dk_ebon_gargoyle : public CreatureScript
                 _initialSelection = false;
                 // Find victim of Summon Gargoyle spell
                 std::list<Unit*> targets;
-                Acore::AnyUnfriendlyUnitInObjectRangeCheck u_check(me, me, 50);
+                Acore::AnyUnfriendlyUnitInObjectRangeCheck u_check(me, me, 30.0f);
```
Comment (UltraNix) [inline src/server/scripts/Pet/pet_dk.cpp:171]: Why lowered that distance? Is it related to this commit at all?

**Review by UltraNix — CHANGES_REQUESTED** [review]

File: src/server/scripts/Pet/pet_dk.cpp (hunk 7 above)
Comment (acidmanifesto) [reply src/server/scripts/Pet/pet_dk.cpp:171]: A error was being thowing up durning the compile testing. I aligned the distance with what tc had and it was shown as a double to float.

File: src/server/game/Grids/Notifiers/GridNotifiers.h (hunk 5 above)
Comment (acidmanifesto) [reply src/server/game/Grids/Notifiers/GridNotifiers.h:806]: This part i just couldnt make sense at all with why it was present. Is there anything in game that would reference this situation of: being alive and indistance and friendly with Not being a Unit or avoid aoe? Or am i misreading that block entirely?

File: src/server/game/Spells/Auras/SpellAuras.cpp (hunk 6 above)
Comment (acidmanifesto) [reply src/server/game/Spells/Auras/SpellAuras.cpp:2813]: I was just not understanding why a z check and los was needed at all here. As auras, even though still being problematic was still being applied to creatures just fine still.

File: src/server/game/Grids/Notifiers/GridNotifiers.h (hunk 5 above)
Comment (UltraNix) [reply src/server/game/Grids/Notifiers/GridNotifiers.h:806]: Being alive, in distance, not friendly and (being a player or not being an aoe avoiding creature). So if creature avoids Aoe, then it should not be put in the list.
It's here for some purpose and I won't remove it cause it might break something.

File: src/server/game/Spells/Auras/SpellAuras.cpp (hunk 6 above)
Comment (UltraNix) [reply src/server/game/Spells/Auras/SpellAuras.cpp:2813]: Because if the target is below or higher than dynamic object owner and the dynamic object owner is not in los with target, the aura should not be applied on target.
It's here for some purpose and I won't remove it unless it breaks something now (but removing this might break something else)

File: src/server/scripts/Pet/pet_dk.cpp (hunk 7 above)
Comment (UltraNix) [reply src/server/scripts/Pet/pet_dk.cpp:171]: I'm talking about `.f` - talking about distance. You changed from 50 yards to 30 yards - why?

Comment (acidmanifesto) [reply src/server/scripts/Pet/pet_dk.cpp:171]: It was throwing a double to float error. Now i could of done it as 50.0f, but this just didnt seem to be consistent with other cores i worked on before. If it needs to be set to 50.0f, then the adjustment wouldnt be a issue in this case to change if you feel like it needs to be set to 50.

File: src/server/game/Spells/Auras/SpellAuras.cpp (hunk 6 above)
Comment (acidmanifesto) [reply src/server/game/Spells/Auras/SpellAuras.cpp:2813]: That would not be a issue if u want it be back in. I just still can not seem to find a scenario or circumstance that actually is utilizing this. Just let me know if you prefer it back in.

File: src/server/game/Grids/Notifiers/GridNotifiers.h (hunk 5 above)
Comment (acidmanifesto) [reply src/server/game/Grids/Notifiers/GridNotifiers.h:806]: Like i mentioned before, im just not able to see where this was put into use. Granted jt seems exceedingly specific, but without a reference for what it is used in particular, im just not certain if this check was still being put into use with as much as the core has evolved. But if you prefer it to be kept that wouldnt be a issue, just let me know what you prefer and the readjustment will be put back in.

File: src/server/scripts/Pet/pet_dk.cpp (hunk 7 above)
Comment (UltraNix) [reply src/server/scripts/Pet/pet_dk.cpp:171]: Yes- change it back to 50.f

File: src/server/game/Spells/Auras/SpellAuras.cpp (hunk 6 above)
Comment (UltraNix) [reply src/server/game/Spells/Auras/SpellAuras.cpp:2813]: Yes - restore that code.

File: src/server/game/Grids/Notifiers/GridNotifiers.h (hunk 5 above)
Comment (UltraNix) [reply src/server/game/Grids/Notifiers/GridNotifiers.h:806]: Yes - restore that code.

**Review by UltraNix — APPROVED** [review]

**Comment by temperrr** [comment]
@acidmanifesto 

How can we test this PR? What do we need to look at specifically for people that are not that good with cpp.

**Comment by acidmanifesto** [comment]
> @acidmanifesto
> 
> How can we test this PR? What do we need to look at specifically for people that are not that good with cpp.

at this point its really just core adjustments for future refactoring. however, 2 additional things were put to be in line with tc which was spellclicks now obey conditions and the core no longer ignores SPELL_ATTR3_ONLY_TARGET_PLAYERS in area auras. However, how to check them in game right now, i am not entirely certain, as this all goes back to my original pr of trying to get aura's working with conditions, i am systematically implementing and refactoring stuff dealing with all parts of the aura system until i can get it working.

File: src/server/game/Grids/Notifiers/GridNotifiers.h [hunk 8]
```diff
@@ -165,6 +165,29 @@ namespace Acore
 
     // WorldObject searchers & workers
 
+        // Generic base class to insert elements into arbitrary containers using push_back
+    template <typename Type> class ContainerInserter
```
Comment (Winfidonarleyan) [inline src/server/game/Grids/Notifiers/GridNotifiers.h:169]:
```suggestion
    // Generic base class to insert elements into arbitrary containers using push_back
    template <typename Type> class ContainerInserter
```

**Review by Winfidonarleyan — APPROVED** [review]

**Comment by temperrr** [comment]
@acidmanifesto We want to test this PR, but all of the testers aren't that familiar with the code. Can you explain in plain english what to look for on this PR, or does this need another dev "tester" to test?

**Comment by acidmanifesto** [comment]
> @acidmanifesto We want to test this PR, but all of the testers aren't that familiar with the code. Can you explain in plain english what to look for on this PR, or does this need another dev "tester" to test?

Honestly it isnt anything players will notice durning game play. So far this is just code style changes/adjustments that will allow easier refactoring with searcher grids which also makes it possible to use vector and dequeue with grid searchers.

The other two adjustments with it were support features because of the changes to not ignore SPELL_ATTR3_ONLY_TARGET_PLAYERS in area auras and fixes Spell Click Conditions to utilize the updated searcher grids. Really this is more of a behind the scene change that players will not see a difference in. 

Tldr: This is something for devs to test and utilized for future refactoring from tc cherry picks and attempt to modernize the code style. Players will not see anything new gameplay wise.

**Comment by Kitzunu** [comment]
If you have tested it then I'm fine with it, per your explination

**Comment by acidmanifesto** [comment]
I have.

**Comment by Winfidonarleyan** [comment]
This pr breaks auras.

**Comment by Nyeriah** [comment]
Confirmed. No auras work at all anymore after this, casting the spells does nothing, auras don't aply.
